### ISSUE TITLE
Ruby: Non-ASCII Method Names Not Recognised

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments#253]
(Original issue 474 created by None on 2010-02-07T23:12:41+00:00)

Ruby 1.9 allows method names to include non-ASCII characters with the following caveats:

* The characters must be valid in the file's source encoding.

* A legal method name that does not end with '!', '?', or '=' may have one of these characters appended.

* The ASCII punctuation characters of which operator methods consist (e.g. {{{[*%&^`~+-/\\[<>=]}}}) must not appear in any other permutation, with the exception of the above case.

Pygments does not recognise such method names, lexing the first non-ASCII character as an error. Examples of unrecognised method names are given in http://pygments.org/demo/3147/ .

Reported by guest

**Comment by Anteru** [comment]
(Original comment by thatch on 2010-05-10T16:40:14+00:00)

I did some digging.  I still can't find a formal announcement, but local rubyers confirm that such support was "rumored."

Checking the source (ruby 1.9 snapshot, `parse.y`) I see some code for this.

{{{
#!c
#define is_identchar(p,e,enc) (rb_enc_isalnum(*p,enc) || (*p) == '_' || !ISASCII(*p))
#define parser_is_identchar() (!parser->eofp && is_identchar((lex_p-1),lex_pend,parser->enc))
...

    mb = ENC_CODERANGE_7BIT;
    do {
        if (!ISASCII(c)) mb = ENC_CODERANGE_UNKNOWN;
        if (tokadd_mbchar(c) == -1) return 0;
        c = nextc();
    } while (parser_is_identchar());
    switch (tok()[0]) {
      case '@': case '$':
        pushback(c);
        break;
      default:
        if ((c == '!' || c == '?') && !peek('=')) {
            tokadd(c);
        }
        else {
            pushback(c);
        }
    }
    tokfix();
}}}

**Comment by Anteru** [comment]
(Original comment by thatch on 2010-05-06T00:37:18+00:00)

Do you have any reference to those rules, or perhaps the grammar itself?  I checked the existing RubyLexer's rules and they're super-complicated:

{{{
            (r'(?:([a-zA-Z_][a-zA-Z0-9_]*)(\\.))?'
             r'([a-zA-Z_][\\w_]*[\\!\\?]?|\\*\\*?|[-+]@?|'
             r'[/%&|^`~]|\\[\\]=?|<<|>>|<=?>|>=?|===?)',
             bygroups(Name.Class, Operator, Name.Function), '#pop'),
}}}

**Comment by Anteru** [comment]
(Original comment by tshatch on 2016-06-02T03:34:01.841778+00:00)

This is still an issue; at tip we wee:
```
#!ruby
def funሴc(x) end # pass, but for the wrong reasons
def sn☃(x) end  # fail
```

**Comment by kurtmckee** [comment]
It appears this is still an issue.

![image](https://user-images.githubusercontent.com/39996/92251347-506a4880-ee92-11ea-9c9e-07772aa36d99.png)

I'll work to fix this so that it follows [the 2.7.0 Ruby docs regarding method names](https://docs.ruby-lang.org/en/2.7.0/syntax/methods_rdoc.html#label-Method+Names).
